tests/e2e-tests/src/server/dot/scenarios.rs

```diff
@@ -0,0 +1,187 @@
+
+/// Test the pdns recursive resolver RFC 9539 opportunistic DoT support works with Hickory.
+///
+/// This tests the scenario where a client recursively queries through a pdns-recursor set up
+/// with DoT probing. After the recursor successfully discovers the Hickory DNS auth servers support
+/// DoT it should start upgrading its requests to use DoT instead of Do53.
+///
+/// We can observe this by making recursive queries and interrogating the traffic that arrives
+/// at the auth server.
+#[test]
+fn pdns_opportunistic_dot_success() -> Result<()> {
+    let network = &Network::new()?;
+    let leaf_ns = NameServer::builder(
+        Implementation::hickory(),
+        FQDN::TEST_DOMAIN,
+        network.clone(),
+    )
+    .pki(Pki::new()?.into())
+    .build()?;
+
+    let Graph {
+        root: root_info,
+        nameservers,
+        ..
+    } = Graph::build(leaf_ns, Sign::No)?;
+
+    // TODO(XXX): Presently auth. hickory servers do not send referrals properly,
+    //   and so only the root NS sees queries. Once #2810 is fixed we should switch
+    //   this test to ensure the other 2 auth. nameservers see expected traffic.
+    let root_ns = nameservers.last().unwrap();
+
+    // Capture nameserver traffic on UDP 53 and TCP 853 to get both plaintext DNS
+    // and DoT requests/responses.
+    let mut tshark = Tshark::builder()
+        .filters(vec![
+            ProtocolFilter::default(),
+            ProtocolFilter::default()
+                .protocol(Protocol::Tcp)
+                .port(DOT_PORT),
+        ])
+        .ssl_keylog_file("/tmp/sslkeys.log") // See hickory.Dockerfile
+        .build(root_ns.container())?;
+
+    // Create a PDNS resolver configured for DoT probing.
+    let resolver = Resolver::new(network, root_info)
+        .custom_config(PDNS_RECURSOR_PROBE_CONFIG.to_string())
+        .start_with_subject(&Implementation::Pdns)?;
+
+    // Make a recursive query for the root FQDN A record using the resolver instance.
+    let client = Client::new(network)?;
+    let query_settings = *DigSettings::default().recurse();
+    let resolver_ip = resolver.ipv4_addr();
+    let resp = client.dig(query_settings, resolver_ip, RecordType::A, &FQDN::ROOT)?;
+    assert!(resp.status.is_noerror());
+
+    // Then, make a second recursive query for a different record type to provoke DoT probing.
+    // The pdns-recursor implementation won't do this based on the first query alone.
+    let resp = client.dig(query_settings, resolver_ip, RecordType::AAAA, &FQDN::ROOT)?;
+    assert!(resp.status.is_noerror());
+
+    // Wait until we've captured at least one incoming DoT probe at the nameserver.
+    tshark.wait_until(
+        |captures| {
+            captures
+                .iter()
+                .filter(|c| {
+                    matches!(c.direction, Direction::Incoming { .. }) && c.dst_port == DOT_PORT
+                })
+                .count()
+                >= 1
```

Comment:
nit: this could be rewritten with `.any()`

Reply:
Good thinking. Fixed :+1: